### PR TITLE
Add the ability to round bar index scale

### PR DESCRIPTION
**Pull request by ericsoco**
I noticed that `ResponsiveBar` charts with many, narrow bars were varying in width. The `svg` container was the correct width, but the bars within it were offset from the left/right edges.

Screencap:
nivo-bar-nicing

After some digging, I narrowed it down to the use of `d3-scale.scaleBand.rangeRound()` within `Bar`'s `common.js::getIndexedScale()`. This diff puts this nicing behavior behind a `nice` prop, which defaults to `true` to maintain current functionality unless a developer opts out with `nice=false`.

The above test with `nice=false`:
nivo-bar-nicing-off

I didn't add anything to the Bar demo site; I can if you like, but this prop feels niche / edge-case enough that it may not need to be surfaced in the demo site.

### CODE REVIEW

**Comment by plouc** [comment]
@ericsoco seems interesting to have the option, however I'm not sure about the naming, having `nice` as a prop on the top level component is hard to figure out I think, maybe something like `roundIndexScale`? This should also be added to the TypeScript definitions, ideally I'd like to have everything documented, but the way the website works is not documented, so I can definitely understand if you don't want to :) It would also be nice to add a unit test for it.

**Comment by wyze** [comment]
I agree about the prop name. Something like `indexScale={{ nice?: boolean, round?: boolean }}` to be more flexible to accept future options if needed.

**Comment by ericsoco** [comment]
@plouc -- I can look into added to TS defs and a unit test; can also take a look at adding to the website if it's not too complex!

Also, I realize now that `nice` is slightly different semantically than `rangeRound` ([docs](https://github.com/d3/d3-scale)), so agreed on naming; can take @wyze's approach and refactor to `indexScale={round}`. I'll leave out "nice" for now, can add in the future if useful.

**Comment by plouc** [comment]
Thank you @ericsoco! Yes @wyze approach is more flexible and align with the `valueScale` we already have.

File: packages/scales/index.d.ts [hunk 1]
```diff
@@ -50,13 +50,19 @@ declare module '@nivo/scales' {
         max?: 'auto' | number
     }
 
+    export interface IndexedScale {
```
Comment (plouc) [inline packages/scales/index.d.ts:53]: This should be a `BandScale` to match d3 types.

File: packages/scales/src/index.js [hunk 2]
```diff
@@ -26,4 +27,5 @@ export const scalePropType = PropTypes.oneOfType([
     PropTypes.shape(timeScalePropTypes),
     PropTypes.shape(logScalePropTypes),
     PropTypes.shape(symLogScalePropTypes),
+    PropTypes.shape(indexedScalePropTypes),
```
Comment (plouc) [inline packages/scales/src/index.js:30]: Would have to be renamed to `bandScalePropTypes`.

File: packages/scales/src/indexedScale.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+ * This file is part of the nivo project.
+ *
+ * Copyright 2016-present, Raphaël Benitte.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+import PropTypes from 'prop-types'
+
+export const indexedScalePropTypes = {
```
Comment (plouc) [inline packages/scales/src/indexedScale.js:11]: This should be `bandScalePropTypes`.

File: packages/bar/index.d.ts [hunk 4]
```diff
@@ -99,6 +99,7 @@ declare module '@nivo/bar' {
         padding: number
 
         valueScale: Scale
+        indexedScale: Scale
```
Comment (plouc) [inline packages/bar/index.d.ts:102]: I think it would be better to call it `indexScale`, the scale type is a bandScale, and is used for indexes.

**Review by plouc — CHANGES_REQUESTED** [review]

**Comment by plouc** [comment]
@ericsoco Thank you for the adjustments, maybe what we suggested was confusing, the property should be `indexScale`, to be consistent with `valueScale` (we cannot really use x/y scale for names as the chart supports different layouts), but the scale type is a band scale.

**Comment by ericsoco** [comment]
No worries, I can make these changes now (and fix the failing lint).

File: packages/bar/src/compute/common.js [hunk 5]
```diff
@@ -15,10 +15,13 @@ import { scaleBand } from 'd3-scale'
  * @param {Function}       getIndex
  * @param {Array.<number>} range
  * @param {number}         padding
+ * @Param {scalePropType}  indexScale
  * @returns {Function}
  */
-export const getIndexedScale = (data, getIndex, range, padding) =>
-    scaleBand().rangeRound(range).domain(data.map(getIndex)).padding(padding)
+export const getIndexedScale = (data, getIndex, range, padding, indexScale) => {
```
Comment (plouc) [inline packages/bar/src/compute/common.js:21]: nit: I think we could also rename this to `getIndexScale()`.

File: packages/bar/src/compute/stacked.js [hunk 6]
```diff
@@ -149,12 +149,13 @@ export const generateStackedBars = ({
...
     padding = 0,
     valueScale,
+    indexScale,
     ...props
 }) => {
     const stackedData = stack().keys(keys).offset(stackOffsetDiverging)(normalizeData(data, keys))
 
     const [axis, range] = layout === 'vertical' ? ['y', [0, width]] : ['x', [height, 0]]
-    const indexedScale = getIndexedScale(data, props.getIndex, range, padding)
+    const indexedScale = getIndexedScale(data, props.getIndex, range, padding, indexScale)
```
Comment (plouc) [inline packages/bar/src/compute/stacked.js:158]: Maybe `indexScale` and `indexScaleConfig`, as you did before?

File: packages/bar/src/props.js [hunk 7]
```diff
@@ -34,6 +34,8 @@ export const BarPropTypes = {
     layout: PropTypes.oneOf(['horizontal', 'vertical']).isRequired,
     reverse: PropTypes.bool.isRequired,
     valueScale: scalePropType.isRequired,
+    indexScale: scalePropType.isRequired,
```
Comment (plouc) [inline packages/bar/src/props.js:37]: Actually, we should use `bandScalePropTypes` as other scale types aren't supported.

File: packages/bar/index.d.ts [hunk 8]
```diff
@@ -99,6 +99,7 @@ declare module '@nivo/bar' {
         padding: number
 
         valueScale: Scale
+        indexScale: Scale
```
Comment (plouc) [inline packages/bar/index.d.ts:102]: We should restrict to BandScale here, otherwise it might give the impression that other types are supported too.

**Review by plouc — CHANGES_REQUESTED** [review]

**Review by wyze — CHANGES_REQUESTED** [review]
Also need to update the canvas implementation to pass the value through to the generate bars methods or it will crash. I made the mistake with valueScale :O.

https://github.com/plouc/nivo/blob/484235ff44bf3cb516e906f8c2b101669b17fc8a/packages/bar/src/BarCanvas.js#L58

File: website/src/pages/bar/api.js [hunk 9]
```diff
@@ -60,6 +60,7 @@ const BarApi = () => {
                     reverse: false,
 
                     valueScale: { type: 'linear' },
+                    indexScale: { type: 'band', round: false },
```
Comment (wyze) [inline website/src/pages/bar/api.js:63]:
```suggestion
                    indexScale: { type: 'band', round: true },
```

File: website/src/pages/bar/canvas.js [hunk 10]
```diff
@@ -41,6 +41,7 @@ const initialProperties = {
     reverse: false,
 
     valueScale: { type: 'linear' },
+    indexScale: { type: 'band', round: false },
```
Comment (wyze) [inline website/src/pages/bar/canvas.js:44]:
```suggestion
    indexScale: { type: 'band', round: true },
```

File: website/src/pages/bar/index.js [hunk 11]
```diff
@@ -39,6 +39,7 @@ const initialProperties = {
     reverse: false,
 
     valueScale: { type: 'linear' },
+    indexScale: { type: 'band', round: false },
```
Comment (wyze) [inline website/src/pages/bar/index.js:42]:
```suggestion
    indexScale: { type: 'band', round: true },
```

**Review by wyze — CHANGES_REQUESTED** [review]
Code looks really good! https://dlqum.sse.codesandbox.io/bar/

Only minor change is I think we should change the website values to default to the default values of the component.

**Comment by ericsoco** [comment]
@wyze oof good catch! Fixed.

**Review by wyze — APPROVED** [review]
Looks great! Thanks. https://z5c1z.sse.codesandbox.io/bar/

**Review by plouc — APPROVED** [review]
LGTM

**Comment by plouc** [comment]
Thank you @ericsoco!

**Comment by ericsoco** [comment]
Thank _you_ @plouc for the lovely library, and @wyze as well 😄

**Comment by jamesdh** [comment]
@ericsoco I'm looking back at this issue and I see your bar chart uses time series data for its X-axis but I don't see any way to do so without converting their value to a string (which causes every bar to have a tick). Can any one of you guys clarify how that was done?

**Comment by ericsoco** [comment]
I have string values for x-axis formatted as `YYYY-MM`, and I'm passing `xFormat: 'time:%Y-%m'` to `ResponsiveBar`.

I don't have a tick on every bar because I'm not using Nivo to render the x-axis labels / ticks, this is a hybrid component that functions as a slider (using Material-UI) with a bar chart overlaid.

I think what I just told you is probably not helpful for your use case...sorry!
